src/System.Buffers.Primitives/System/Internal/ManagedBufferPool.cs

```diff
@@ -70,7 +80,17 @@ public override BufferHandle Pin(int index = 0)
                 }
             }
 
-            byte[] _array;
+            public override void Retain()
+            {
+                if (IsDisposed) BufferPrimitivesThrowHelper.ThrowObjectDisposedException(nameof(ArrayPoolBuffer));
+                Interlocked.Increment(ref _referenceCount);
+            }
+
+            public override void Release()
+            {
+                Debug.Assert(!IsDisposed);
```

Comment:
The ref count will possibly get below zero. I could do a release check, but it does not really address the problem, i.e. an invalid double release could suceed, and then valid single release would throw, e.g.
```c#
void Bar() {
    OwnedBuffer<byte> buffer = ...
    buffer.Retain();
    Foo(buffer);
    buffer.Release(); // this will throw despite the fact that this method is correct 
}
void Foo(OwnedBuffer<T> buffer){
    buffer.Retain();
    buffer.Relase();
    buffer.Release();
}
```